compiler/AST/Thunk.cpp

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2020-2023 Hewlett Packard Enterprise Development LP
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Thunk.h"
+
+#include "AstVisitor.h"
+#include "global-ast-vecs.h"
+
+Thunk::Thunk() : Expr(E_Thunk) {
+  children.parent = this;
+  gThunks.push_back(this);
+}
+
+void Thunk::verify() {
+}
+
+void Thunk::accept(AstVisitor* visitor) {
+  if (visitor->enterThunk(this)) {
+    for_alist(child, children) {
+      child->accept(visitor);
+    }
+    visitor->exitThunk(this);
+  }
+}
+
+Thunk* Thunk::copyInner(SymbolMap* map) {
+  INT_FATAL("thunks should not be copied");
```

Comment:
Hmm, I wonder if you could end up with this, with nested coforalls or something?

Reply:
My thinking was that copying is deferred until thunks are forced, and that thunks would be forced in a "deeper-first" order (because we convert the AST "bottom up"). This would mean that by the time a thunk is forced, it should contain no inner thunks. I'll check that nested coforall cases don't break, though.

Reply:
The following works fine under `--dyno --verify`:
```Chapel
class C {}
coforall 1..10 {
    coforall 1..10 {
        if var x = new C() {
            x;
        }
    }
}
```